contrib_bots/lib/gcalendar.py

```diff
@@ -0,0 +1,263 @@
+#
+# The ideal situation will be when Zulip becomes capable of managing API
+# keys and user credentials, allowing the user to choose between which
+# bots have access to what. However, this belongs more to the long-term
+# plan in Zulip's roadmap.
+
+def get_credentials():
+    '''Gets valid user credentials from storage.
+
+    If nothing has been stored, or if the stored credentials are invalid,
+    the OAuth2 flow is completed to obtain the new credentials.
+
+    Returns:
+        Credentials, the obtained credential.
+    '''
+    home_dir = os.path.expanduser('~')
+    credential_dir = os.path.join(home_dir, '.gcalbot-credentials')
+    if not os.path.exists(credential_dir):
+        os.makedirs(credential_dir)
+    credential_path = os.path.join(credential_dir,
+                                   'gcalbot.json')
+
+    store = Storage(credential_path)
+    cred = store.get()
+    if not cred or cred.invalid:
+        flow = client.flow_from_clientsecrets(os.path.join(home_dir, CLIENT_SECRET_FILE), SCOPES)
+        flow.user_agent = APPLICATION_NAME
+        flags = argparser.parse_args('--noauth_local_webserver --logging_level INFO'.split())
+        cred = tools.run_flow(flow, store, flags)
+    return cred
+
+def reply_message(msg, original_msg, client):
+    '''Replies to a specific message, through the channel of origin.
+
+    Keyword arguments:
+    msg -- content of the reply
+    original_msg -- message that is being replied
+    client -- client from the message handler
+    '''
+    message_type = original_msg['type']
+
+    if message_type == 'private':
+        original_sender = original_msg['sender_email']
+
+        client.send_message(dict(
+            type=message_type,
+            to=original_sender,
+            content=msg,
+        ))
+    elif message_type == 'stream':
+        original_stream = original_msg['display_recipient']
+        original_subject = original_msg['subject']
+
+        client.send_message(dict(
+            type=message_type,
+            to=original_stream,
+            subject=original_subject,
+            content=msg,
+        ))
+    else:
+        raise UnknownMessageTypeError(message_type)
+
+'''Identifies and parses the different parts of a message sent to the bot
+
+Returns:
+    Values, a tuple that contains the 2 (or 3) parameters of the message
+'''
+def parse_message(message):
+    content = message.replace('@gcalendar ', '')
+
+    try:
+        splits = content.split('|')
+        title = splits[0].strip()
+        start_date = dateparser.parse(splits[1], settings={'RETURN_AS_TIMEZONE_AWARE': True})
+        end_date = None
+
+        if len(splits) == 3: # Not a full-day event
+            end_date = dateparser.parse(splits[2], settings={'RETURN_AS_TIMEZONE_AWARE': True})
+
+    except Exception as e:
+        logging.error(e)
+
+        return None, None, None
+
+    return title, start_date, end_date
+
+credentials = get_credentials()
```

Comment:
Just moved `credentials`, `http` and `service` inside the handler class, so this should limit the scope of these variables to a more adequate one. See my new commit for details.